### ISSUE TITLE
Memory leak in tf.math.reduce_sum 

### ISSUE THREAD

**Issue by bhoeveler** [ROCmSoftwarePlatform/tensorflow-upstream#1149]
**System information**

- Linux Opensuse 15.2 Kernel 5.9.1
- Tensorflow 2.3.0 from pip
- Radeon VII  with rocm 3.8 
- Python 3.6.10

**Describe the current behavior**
The current implementation of tf.math.reduce_sum with the 64bit datatype will leak memory on my machine.
Minimal example:

```python
res = tf.zeros((256,256), dtype=tf.float64)
for i in range( 1000000):
     A = tf.random.normal(( 6,6,256,256 ),dtype=tf.float64)
     r = tf.math.reduce_sum( A, axis=(0,1))
     res += r
     del A, r
     if i % 1000 == 0:
          gc.collect()
     print(i)
```
After the loop around 4 GB Ram are not accessible anymore. I have no idea which log-files might be useful, just tell me.

Edit:
The memory leak appears in GEMM and transpose too, maybe that's the reason. However it makes it hard to find a workaround.

**Comment by jerryyin** [comment]
Thanks for bring this up. A few clarifications to make sure we are on the same page: 
 - When saying that 4GB RAM are not accessible, do you mean:
   - It is GPU or CPU memory? How do you determine they are not accessible?
   - Does it occur with only float64 type?
- I don't think we formally support Opensuse, is it possible to give Ubuntu based docker a shot? `docker pull rocm/tensorflow:rocm3.8-tf2.3-dev`

**Comment by bhoeveler** [comment]
>When saying that 4GB RAM are not accessible, do you mean:
>   It is GPU or CPU memory? How do you determine they are not accessible?

I ment system memory or CPU memory. When I'm talking about not accessible I mean by any python-variable, since I deleted all of them and ran the garbage-collector. However Valgrind indicates, that there are pointers left that are not used anymore -- See below. GPU-memory is not an issue. If I increase the number of iterations I can provoke a crash by out-of-memory.

> Does it occur with only float64 type?

No, I didn't checked at the beginning since it wasn't relevant for me, but apparently it still occurs for float32 too. After the loop python uses 3.9 GByte of Ram, which is slightly less than in the float64 case. 

> I don't think we formally support Opensuse, is it possible to give Ubuntu based docker a shot? docker pull rocm/tensorflow:rocm3.8-tf2.3-dev

I used the rpm's for SLES 15 SP 2 which should not differ too much from Opensuse 15.2 when it comes to the relevant core-libraries. Now, I also ran the docker-image which shows the same behavior. I think the leak appears in tensorflow-rocm and not rocm itself. The issue still exits with Rocm 3.9 and tensorflow 2.3.2. Can you reproduce the issue ?

Where I think the leak appears as reported by Valgrind, but I don't have the knowledge to persuade it further:

```
==31785== 4,480,000 bytes in 10,000 blocks are still reachable in loss record 147,711 of 147,714
==31785==    at 0x4C2E94F: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==31785==    by 0x508B0FA5: ??? (in /opt/rocm-3.9.0/hip/lib/libamdhip64.so.3.9.30900)
==31785==    by 0x508C9A45: ??? (in /opt/rocm-3.9.0/hip/lib/libamdhip64.so.3.9.30900)
==31785==    by 0x508ADDB2: hipLaunchKernel (in /opt/rocm-3.9.0/hip/lib/libamdhip64.so.3.9.30900)
==31785==    by 0x665E3224: Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<double, 1, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, int>, 16, Eigen::MakePointer> const, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::GpuDevice, true, (Eigen::internal::TiledEvaluation)0>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<double, 1, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, int>, 16, Eigen::MakePointer> const, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const&, Eigen::GpuDevice const&) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x665DB5BB: tensorflow::functor::BinaryFunctor<Eigen::GpuDevice, tensorflow::functor::add<double>, 1, false>::operator()(Eigen::GpuDevice const&, Eigen::TensorMap<Eigen::Tensor<double, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<double const, 1, 1, long>, 16, Eigen::MakePointer>, bool*) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x65902272: tensorflow::BinaryOp<Eigen::GpuDevice, tensorflow::functor::add<double> >::Compute(tensorflow::OpKernelContext*) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x4F4C3972: tensorflow::BaseGPUDevice::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/libtensorflow_framework.so.2)
==31785==    by 0x5DBA5486: tensorflow::KernelAndDeviceOp::Run(tensorflow::ScopedStepContainer*, tensorflow::EagerKernelArgs const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::CancellationManager*, absl::lts_2020_02_25::optional<tensorflow::EagerRemoteFunctionParams> const&) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x5DB6CC9E: tensorflow::EagerKernelExecute(tensorflow::EagerContext*, absl::lts_2020_02_25::InlinedVector<tensorflow::TensorHandle*, 4ul, std::allocator<tensorflow::TensorHandle*> > const&, absl::lts_2020_02_25::optional<tensorflow::EagerRemoteFunctionParams> const&, std::unique_ptr<tensorflow::KernelAndDevice, tensorflow::core::RefCountDeleter> const&, tensorflow::GraphCollector*, tensorflow::CancellationManager*, absl::lts_2020_02_25::Span<tensorflow::TensorHandle*>) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x5DB6D90A: tensorflow::ExecuteNode::Run() (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
==31785==    by 0x5DB9F3E9: tensorflow::EagerExecutor::SyncExecute(tensorflow::EagerNode*) (in /home/bhoeveler/.virtualenv/develop-tf/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
```

**Comment by rsanthanam-amd** [comment]
A memory leak has been found in HIP which is causing this issue.  A ticket has been filed with the HIP team to address it.

When a fix is available, it will be verified prior to closing out this ticket.

**Comment by bhoeveler** [comment]
Will the fix find its way into next release ? Or can you name an approximate date ?

**Comment by rsanthanam-amd** [comment]
Upon further research, we discovered that the memory does actually get released but only if the tensorflow runtime is synchronized periodically.

Here is one suggested way of doing that in the minimal example:

```import tensorflow as tf
res = tf.zeros((256,256), dtype=tf.float64)
for i in range( 1000000):
     A = tf.random.normal(( 6,6,256,256 ),dtype=tf.float64)
     r = tf.math.reduce_sum( A, axis=(0,1))
     res += r
     del A, r
     if i % 10000 == 0:
          print(i,res)
```

**Comment by bhoeveler** [comment]
Thanks ! I was able to write a workaround for my original code ( simulation ). However, I think it has a impact performance-wise ( I did not sync it periodically but every time, since it's very hard to integrate it into the code otherwise ).

I used 
```python
res = tf.zeros((256,256), dtype=tf.float64)
for i in range( 1000000):
     A = tf.random.normal(( 6,6,256,256 ),dtype=tf.float64)
     r = tf.math.reduce_sum( A, axis=(0,1))
     res += r
     del A, r
     if i % 10000 == 0:
          _ = res.numpy()
```
instead, which is a bit faster. I would like to see a fix, tho.

Edit: Ok, it fixed one leak, but still leaking - only far less and will probably crash after a few hours.

**Comment by rsanthanam-amd** [comment]
An easier way to synchronize every time is to create the following environment variable:

AMD_SERIALIZE_KERNEL=3

This will eliminate the need for user code changes to accomplish this.

Based on the current minimal example, I have not observed any other leaks.

Please provide another minimal example which exhibits the additional leak.

**Comment by bhoeveler** [comment]
I rerun it with the enviroment variable set, but tensorflow is still leaking memory. I will look into it and will provide a minimal example.

Any other variables that may be relevant and worth a shot ?

**Comment by bhoeveler** [comment]
```Python
import tensorflow as tf
import numpy as np

def GMRes(x0, b , linOp ,
             inner = lambda x,y : tf.math.reduce_sum( x * y ).numpy(),
             norm = lambda x : tf.linalg.norm(x).numpy(),
             zeros_like = lambda x : tf.zeros_like(x),
             maxIters = [10,20,30 ],maxRestarts = 100, eps = 1e-5, delta = 1e-5):
    
    if norm is None:
        norm = lambda x : np.sqrt( inner( x,x ) )
    x = x0
    nb = norm( b )
    for restarts in range(maxRestarts):
        print( "GMRes restarted :", restarts)
        r = b - linOp( x )
        nr = norm( r ) 
        print( "current residual: ", nr / nb)
        if nr / nb < eps:
            return x
        if restarts < len( maxIters )  :            
            maxIter = maxIters[restarts]
        else:
            maxIter = maxIters[-1]
        q = [0] * (maxIter)
    
    
        q[0] = r / nr
    
        h = np.zeros((maxIter + 1, maxIter))
    
        for k in range(maxIter):
            y = linOp( q[k] )
            for j in range(k + 1):
                h[j, k] = inner( q[j], y )
                y = y - h[j, k] * q[j]
                
            h[k + 1, k] = norm( y ) 
            if (h[k + 1, k] != 0 and k != maxIter - 1):
                q[k + 1] = y * ( 1. / h[k + 1, k] )
            else:
                break
    
        d = np.zeros(k + 1)
        d[0] = nr
        
        result = np.linalg.lstsq(h[:k+1,:k], d)[0]
        h = sum( (qi * ri for qi, ri in zip(q,result)) ) 
        nx = norm(x)
        if norm( h ) / nx < delta:
            return x + h 
        x = x + h
        del q
        
    print( "konvergence failed -- returning non-optimal value")
    return x

n = 2048
dtype= tf.float64
delta = 2e-2

res = tf.zeros((n,), dtype=dtype)
for i in range(10000):
    b = tf.random.normal((n,), dtype=dtype)
    A = tf.linalg.eye(n,dtype=dtype)  + delta * tf.random.normal((n,n),dtype=dtype)
    linOp = lambda x : tf.linalg.matvec( A, x)
    res += GMRes(b,b,linOp)
```

I couldn't condense it further, my original code uses GMRes too, but in a bit more advanced setting. This example will leak memory on my machine - but very slowly - even with AMD_SERIALIZE_KERNEL=3. On my machine the simulation crashes after ~ 1 h without reaching the first checkpoint.

**Comment by ekuznetsov139** [comment]
The problem here is that you're using Tensorflow in eager mode, and keep bouncing between Tensorflow and numpy. It's not intended to be used this way for long periods of time. (And you can't switch to graph mode because of all references to numpy.) Tensorflow caches scalar objects when in eager mode, and, since eager mode is generally meant for short runs, never bothers to empty that cache. Your code continuously adds more and more objects to the cache and ends up using a noticeable amount of memory there.

It takes a bit of work to rewrite the script in such way that it can be run in graph mode, but here's my best shot:
```
import tensorflow as tf
import numpy as np
import time
dtype= tf.float64

def inner_inner_loop(q, k, y, inner):
    row = tf.TensorArray(dtype, size=k)
    def body_op(row, j, y):
        qj = q.read(j)
        temp = inner( qj, y )
        y = y - temp * qj
        row = row.write(j, temp)
        return row, j+1, y
    def cond_op(row, j, y):
        return j<k
    j = tf.constant(0)
    row, _, y = tf.while_loop(cond_op, body_op, [row, j, y])
    return row.stack(), y

def inner_inner_loop_v2(q, k, y, inner):
    qs=q.stack()
    inners = tf.math.reduce_sum(qs*tf.reshape(y,[1,-1]), axis=1)
    product = tf.reshape(inners,[-1,1])*qs
    delta = tf.reduce_sum(product, axis=0)
    return inners[:k], y-delta

def inner_loop(maxIter, q0, linOp, norm, inner):
        h = tf.TensorArray(dtype, size=maxIter)
        q = tf.TensorArray(dtype, size=maxIter+1,clear_after_read=False)
        q = q.write(0,q0[:])
        temp = tf.reshape(tf.constant(1.0,dtype=dtype),[1])
        k = tf.constant(0)
        def cond_op(q_prev, k, q, h, temp):
          return temp!=0
        def body_op(q_prev, k, q, h, temp):
            y = linOp( q_prev )
            row, y = inner_inner_loop(q, k+1, y, inner)
            #row, y = inner_inner_loop_v2(q, k+1, y, inner)
            temp = norm( y )
            temp = tf.reshape(temp, [1])
            row = tf.concat([row,temp], axis=0)
            row = tf.concat([row, tf.zeros((maxIter-k), dtype=dtype)], axis=0)
            h = h.write(k, row)
            q_prev = tf.cond(temp==0, lambda:q_prev, lambda:y / temp)
            q = q.write(k+1, q_prev)
            return tf.reshape(q_prev,[-1]), k+1, q, h, temp
        _, k, q, h, _ = tf.while_loop(cond_op, body_op, [q0, k, q, h, temp], maximum_iterations=maxIter,
                shape_invariants=[q0.get_shape(), k.get_shape(),
                tf.TensorShape([]),
                tf.TensorShape([]),
                temp.get_shape()])
        h=tf.transpose(h.stack(),[1,0])
        q=tf.transpose(q.stack(),[1,0])
        return q, h, k

@tf.function
def GMRes(x0, b, linOp = lambda x : tf.linalg.matvec( A, x),
             inner = lambda x,y : tf.math.reduce_sum( x * y ),
             norm = lambda x : tf.linalg.norm(x),
             zeros_like = lambda x : tf.zeros_like(x),
             maxIters = [10,20,30 ], maxRestarts = 100, eps = 1e-5, delta = 1e-5):
    if norm is None:
        norm = lambda x : tf.math.sqrt( inner( x,x ) )
    x = x0
    nb = norm( b )
    r = b - linOp( x )
    nr = norm(r)
    h_x = tf.constant(2*delta,dtype=dtype)
    maxIterLen = len(maxIters)-1
    maxIters = tf.constant(maxIters)
    def cond_op(x, r, nr, h_x, restarts):
        return tf.math.logical_and(nr/nb>=eps, h_x>=delta)
    def body_op(x, r, nr, h_x, restarts):
        with tf.control_dependencies([tf.print( "current residual: ", nr / nb)]):
          maxIter = tf.gather(maxIters, tf.minimum(restarts, maxIterLen-1))
        q, h, k = inner_loop(maxIter, r/nr, linOp, norm, inner)
        d = tf.reshape(tf.concat([tf.reshape(nr,[1]), tf.zeros(k,dtype=dtype)], axis=0),[-1,1])
        result = tf.linalg.lstsq(h[:k+1,:k], d,fast=False)[:,0]
        h = tf.reduce_sum(q[:,:-1]*result, axis=1)
        x = x + h
        r = b - linOp(x)
        nr = norm(r)
        restarts = restarts + 1
        return x, r, nr, norm(h)/norm(x), restarts
    
    restarts = tf.constant(0, dtype=tf.int32)
    x, _, _, _, _ = tf.while_loop(cond_op, body_op, (x, r, nr, h_x, restarts), maximum_iterations=maxRestarts)
    
#    print( "konvergence failed -- returning non-optimal value")
    return x
n = 2048
delta = 2e-2
res = tf.zeros((n,), dtype=dtype)
t=time.time()
for i in range(10000):
    print("***", i, "***")
    b = tf.random.normal((n,), dtype=dtype)
    A = tf.linalg.eye(n,dtype=dtype)  + delta * tf.random.normal((n,n),dtype=dtype)
    res += GMRes(b,b)
```

No memory leak and it runs faster than the original. (It seems to work, but I can't guarantee 100% correctness. Also note inner_inner_loop_v2: it assumes a specific form of "inner", but speeds up the code even further.)

The memory leak is not AMD-specific and you could try to raise the issue in the upstream repository (/tensorflow/tensorflow), though i would not give good odds for it being fixed there.

**Comment by bhoeveler** [comment]
Thank you very much for your help and time, even it wasn't an amd specific error. I was just checking against the CPU-device which didn't had the first "leak" and was too slow to show the second and then I did blame rocm. 

I'll integrate your code since it is faster and the use of tf.function is generally a good thing. Thanks to your answer I also found a fix for my messy code by clearing the caches manually in each restart:

```python
...
from tensorflow.python.eager import context

def GMRES(....):
   for restarts in range(maxRestarts):
     ...
     context.context()._clear_caches()
```

But I think I'll stick to your code since it's just better practice and even faster. I close the issue now, since it's not amd or tensorflow related. Thank you very much for our help and support once again.
